llext: add threads and kernel object test case

laurenmurphyx64 opened the pull request:
Tests threads and kernel objects with userspace.

laurenmurphyx64 committed:
llext: add threads and kernel object test case

Tests threads and kernel objects with userspace. Includes cleanups
suggested by Luca Burelli.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

pillo79 commented on tests/subsys/llext/simple/src/threads_kernel_objects_ext.c:
```diff
@@ -0,0 +1,68 @@
+#include <zephyr/kernel.h>
+
+#ifdef CONFIG_USERSPACE
+extern struct k_sem my_sem_userspace;
+extern struct k_thread my_thread_userspace;
```
I would (re)use the local names even in the userspace case. With a few extra defines such as `MY_THREAD_PRIO`, `MY_THREAD_FLAGS` etc, most if not all of the other `#ifdefs` could go away :slightly_smiling_face: 
These `extern`s should also probably join other things in the shared .h for type safety.

laurenmurphyx64 replied:
Great suggestion, it's pretty messy as-is. Have put them in a header.

pillo79 commented on tests/subsys/llext/simple/src/threads_kernel_objects_ext.c:
```diff
@@ -0,0 +1,68 @@
+static struct k_sem my_sem;
+#endif
+
+extern k_thread_stack_t my_thread_stack;
+#define STACK_SIZE 1024
```
Move to .h

pillo79 reviewed (commented):
Nice additional test, thank you! Some minor feedback on this.

laurenmurphyx64 commented on tests/subsys/llext/simple/src/test_llext_simple.c:
```diff
@@ -50,7 +51,25 @@ int z_impl_ext_syscall_ok(int a)
+EXPORT_SYMBOL(my_thread_stack);
+
 #ifdef CONFIG_USERSPACE
+K_SEM_DEFINE(my_sem_userspace, 1, 1);
```
I think I fixed it, let me know if I misunderstood

laurenmurphyx64 commented:
The compliance check on this one is destined to fail because checkpatch can't detect I actually did export the variable after I declared it:

```
ERROR   : Test Checkpatch failed: 
EXPORT_SYMBOL: EXPORT_SYMBOL(foo); should immediately follow its function/variable
File:tests/subsys/llext/simple/src/test_llext_simple.c
Line:59
```

Lines 58 and 59:
```
K_THREAD_STACK_DEFINE(my_thread_stack, MY_THREAD_STACK_SIZE);
EXPORT_SYMBOL(my_thread_stack);
```

@teburd Should we add exports to the checkpatch ignore, or just let this one pass, or am I missing something...?

teburd commented:
we should tell checkpatch to ignore EXPORT_SYMBOL imho

laurenmurphyx64 committed:
checkpatch: ignore EXPORT_SYMBOL

checkpatch isn't smart enough to detect a function / variable has
been declared e.g. with K_SEM_DEFINE and will incorrectly flag
such cases.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

laurenmurphyx64 commented:
> we should tell checkpatch to ignore EXPORT_SYMBOL imho

@teburd Added a commit to do that, others welcome to weigh in on it

pillo79 reviewed (commented):
Thanks, this looks way better - basically just one comment from me.
 
I was still confused by all the remaining `#ifdef`s though, so I played around with the sample a bit. The result are two extra commits in my repo [here](https://github.com/pillo79/zephyr/tree/threads_objects_llext_sample); please take a look and feel free to integrate any bits or pieces from there in this PR if you think they are useful :slightly_smiling_face:

laurenmurphyx64 commented on tests/subsys/llext/simple/src/test_llext_simple.c:
```diff
@@ -7,9 +7,11 @@
 #include <zephyr/logging/log.h>
 #include <zephyr/sys/libc-hooks.h>
+#include "threads_kernel_objects.h"
 
```
Thanks! I took your commits and squashed them into my original + changed the name of threads_kernel_objects_ext.h / moved it below syscalls_ext.h

pillo79 reviewed (approved):
Love it!
